client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java

```diff
@@ -63,17 +63,19 @@ public static Iterable<List<Object>> fixData(List<Column> columns, List<List<Obj
         }
         ColumnTypeHandler[] typeHandlers = createTypeHandlers(columns);
         ImmutableList.Builder<List<Object>> rows = ImmutableList.builderWithExpectedSize(data.size());
-        for (List<Object> row : data) {
-            if (row.size() != typeHandlers.length) {
-                throw new IllegalArgumentException("row/column size mismatch");
-            }
-            for (int i = 0; i < row.size(); i++) {
-                Object value = row.get(i);
-                if (value != null) {
-                    row.set(i, typeHandlers[i].fixValue(value)); // update value in-place
+        try {
+            for (List<Object> row : data) {
+                for (int i = 0; i < row.size(); i++) {
+                    Object value = row.get(i);
+                    if (value != null) {
+                        row.set(i, typeHandlers[i].fixValue(value)); // update value in-place
+                    }
                 }
+                rows.add(unmodifiableList(row)); // allow nulls in list
             }
-            rows.add(unmodifiableList(row)); // allow nulls in list
+        }
+        catch (IndexOutOfBoundsException e) {
+            throw new IllegalArgumentException("row/column size mismatch");
```

Comment:
This will only catch the case when we have for objects in row than types. Not the other way around.